### DISCUSSION TITLE
Apollo Java: how do i wait for a response using Rx3Apollo

### DISCUSSION THREAD

**Discussion by JustRed23** [apollographql/apollo-server#5777] [Q&A] [answered]
My code for now is this: 

```java
    public Result<Status, T> execute() {
        AtomicReference<Result<Status, T>> result = new AtomicReference<>(Result.result(Status.FAILED, null));

        Rx3Apollo.from(getGraphQLCall())
                .filter(response -> !response.hasErrors())
                .map(Response::getData)
                .subscribeOn(Schedulers.io())
                .observeOn(Schedulers.computation())
                .blockingSubscribe(successfulResponse -> result.set(Result.result(Status.SUCCESS, successfulResponse)), error -> {
                    if (error instanceof ApolloNetworkException networkException)
                        result.set(Result.result(Status.NO_NETWORK.setThrowable(networkException), null));
                    result.set(Result.result(Status.FAILED.setThrowable(error), null));
                });
        return result.get();
    }
```

This works fine but when this is finished the client does not exit. I looked at the threads and it shows that this call is still running. How do i fix this?

**Comment by martinbonnin** [comment, marked as answer]
Hi 👋  ! Looks like this is an apollo-android issue? I'll move this discussion to an issue there. First time I try this in production 😅  . Finger crossed it'll land in the correct repo with all the comments :)

**Reply by martinbonnin** [reply]
@JustRed23 Turns out that moving a discussion to an issue in a different repo will lose both the original author and comments. Do you mind creating an issue in apollo-android at https://github.com/apollographql/apollo-android/issues/new?assignees=&labels=Type%3A+Question&template=question.md&title= ? 

And we'll take it from there. Thanks!

**Reply by JustRed23** [reply]
To be honest... I kind of forgot about this thread I made. I am using another library right now since apollo does'nt really support Java anymore.

**Reply by martinbonnin** [reply]
Apollo-Android supports Java. See https://www.apollographql.com/docs/android/essentials/get-started-java/ for an intro. I'm curious what makes you think otherwise? 

Regarding this specific issue, I suspect it has to do with OkHttp threadpools being left open, which you can close by calling [shutdown()](https://square.github.io/okhttp/4.x/okhttp/okhttp3/-ok-http-client/#shutdown-isnt-necessary). 

In all cases, I'm happy to hear you found a solution. If you ever have any question regarding Java support or anything Apollo Android in general, feel free to open an issue in the repo: https://github.com/apollographql/apollo-android/issues/new
